### ISSUE TITLE
Y24-427 - As a team (PSD) we would like to research, define and agree a branching strategy so that untested features in develop do not impede deployment of other features.

### ISSUE THREAD

**Issue by TWJW-SANGER** [sanger/General-Backlog-Items#463]
**Description**
As a team (PSD) we would like to research, define and agree a branching strategy so that untested features in develop do not impede deployment of other features.

**Acceptance Criteria**
The following outputs from the research are produced:
- [ ] A strategy that allows multiple features to be worked on simultaneously such that feature/s in UAT don't impact the testability of each other.
- [ ] A strategy that allows deployments to production of only _tested_ features ready for deployment.
- [ ] A confluence page in the Confluence page PSD Public - Public Processes describing the branching strategy - agreed by consensus.

**Who the primary contacts are for this work**
PSD

**Additional context or information**
Currently process is:
* Code goes into develop, then it gets tested.  
* Merge develop into master (with all the unapproved stuff that might be in there).
* Release master to production.
 
Suggestion is that we create an additional WIP branch for cutting edge features. 
We need to work through how this will be used in the case of multiple features being worked on in parallel that will be potentially tested at different points in time by different end users.
Ideally we can deploy the different features to UAT and Production in any combination with minimal complex manual unpicking.....

**Comment by BenTopping** [comment]
https://docs.google.com/document/d/1LqtkTv817ju3yMuZM5p5dOUdKdT5_5bmjkgNdIIXVz4/edit?usp=sharing

**Comment by harrietc52** [comment]
Nice work! 👍 

- I like the proposal to deploy to UAT via feature branch, and only merge into `develop` when the feature is approved. I think that a simple and effective approach 
- Also like the idea of scheduling UAT time, but it might be worth speaking with the customers to see how they would feel if we did this
- Like you say, there would be more manual steps involved in creating a release manually from a feature branch, and the time it takes for the release to build. But its not a complex step, as long as it is well documented. 
- The other thing I wondered, is we might need to be able to specify in the intergration suite what feature branch we want it to run against, becuase if it is pointing to `develop`, that would be too late if there were issues. This might already be possible, I wasn’t sure 

I feel like this is a great start, and we should _just_ give it a go 😄  learn from it, then we can keep coming back to this original proposal if we feel like more is needed

**Comment by KatyTaylor** [comment]
I've added a bunch of suggestions to your google doc but they are mostly small things. Basically I think it all sounds very reasonable, and clearly communicated, thank you.

@harrietc52 - when testing with the Integration Suite against UAT, you can specify which Integration Suite branch you want to use. If that's what you were asking?

**Comment by BenTopping** [comment]
Closing as research is complete.
- https://github.com/sanger/developer-laptop-setup/pull/25
- https://ssg-confluence.internal.sanger.ac.uk/display/PSDPUB/Process+for+Git+Branching+and+UAT+testing
- https://docs.google.com/document/d/1LqtkTv817ju3yMuZM5p5dOUdKdT5_5bmjkgNdIIXVz4/edit?usp=sharing
